### ISSUE TITLE
Not conditionally typed inside match statement

### ISSUE THREAD

**Issue by DeviousStoat** [microsoft/pyright#2628]
**Describe the bug**
A generic type isn't conditionally typed when type matching inside a match statement.

**To Reproduce**
```python
from typing import TypeVar

_T = TypeVar("_T")

def f(v: _T) -> _T | None:
    match v:
        case str():
            reveal_type(v)  # Type of "v" is "str" (should probably be "str*")
            return v  # error: Expression of type "str" cannot be assigned to return type "_T@f | None"
        case _:
            return None
```

**Expected behavior**
Same as this:
```python
from typing import TypeVar

_T = TypeVar("_T")

def f(v: _T) -> _T | None:
    if isinstance(v, str):
        reveal_type(v)  # Type of "v" is "str*"
        return v # No error
    return None
```

**VS Code extension or command-line**
```bash
$ pyright --version
pyright 1.1.190
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.191, which I just published. It will also be included in the next release of pylance.
